# Credit Risk Evaluator

In [11]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
from pathlib import Path
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler

## Retrieve the Data

The data is located in the Challenge Files Folder:

* `lending_data.csv`

Import the data using Pandas. Display the resulting dataframe to confirm the import was successful.

In [12]:
# Import the data
lending_data = pd.read_csv("./Resources/lending_data.csv")

# Preview dataframe
lending_data.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [13]:
# Check for rows with missing values
lending_data.isnull().sum()

loan_size           0
interest_rate       0
borrower_income     0
debt_to_income      0
num_of_accounts     0
derogatory_marks    0
total_debt          0
loan_status         0
dtype: int64

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

##### My Prediction

I predict that the logistic regression will perform better since there are only 7 features and all seem like they could plausibly be related to whether a loan would be denied or approved (i.e. less noise in the parameters to tune out). We are only categorizing the data into two possible groups (approved/denied) so it seems like logistic regression would be the better choice.

## Split the Data into Training and Testing Sets

In [14]:
# Split the data into X_train, X_test, y_train, y_test
X = lending_data.drop(columns=["loan_status"])
y = lending_data["loan_status"]

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)


## Create, Fit and Compare Models

Create a Logistic Regression model, fit it to the data, and print the model's score. Do the same for a Random Forest Classifier. You may choose any starting hyperparameters you like. 

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the designated markdown cell.

In [28]:
# Create logistic regression model and fit to training data
model_lr = LogisticRegression().fit(X_train, y_train)

# Print training and testing accuracy scores
print(f'Training Score: {model_lr.score(X_train, y_train)}')
print(f'Testing Score: {model_lr.score(X_test, y_test)}')

Training Score: 0.9921240885954051
Testing Score: 0.9918489475856377


In [17]:
# Scale training and testing data based on training features
scaler = StandardScaler().fit(X_train)

# Apply scaling transformation to training/testing features
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [30]:
# Train a Random Forest Classifier model and print the model score
model_rfc = RandomForestClassifier(random_state=1, n_estimators=100).fit(X_train_scaled, y_train)
print(f'Training Score: {model_rfc.score(X_train_scaled, y_train)}')
print(f'Testing Score: {model_rfc.score(X_test_scaled, y_test)}')

Training Score: 0.9975409272252029
Testing Score: 0.9914878250103177


The logistic regression slightly outperformed the random forest model in accuracy as well as in computational speed. The logistic regression had a score of about 99.18% on the testing data while the random forest classifier had a score of about 99.15%. This barely matched my original prediction, although the random forest classifier took much longer. The logistic regression took 0.1 seconds (or less) to fit and score while the random forest classifier took over two seconds. With a larger dataset the logistic regression would certainly be the optimal choice.